ql/src/java/org/apache/hadoop/hive/ql/optimizer/metainfo/annotation/OpTraitsRulesProcFactory.java

```diff
@@ -158,8 +158,57 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
         }
       }
 
-      listBucketCols.add(bucketCols);
-      OpTraits opTraits = new OpTraits(listBucketCols, numBuckets,
+      final List<List<String>> listBucketCols = new ArrayList<>();
+      final List<CustomBucketFunction> bucketFunctions = new ArrayList<>();
+      final int numBuckets;
+      if (parentOpTraits == null || !parentOpTraits.hasCustomBucketFunction()) {
+        // No CustomBucketFunctions
+        listBucketCols.add(bucketCols);
+        bucketFunctions.add(null);
+        numBuckets = parentOpTraits == null ? -1 : parentOpTraits.getNumBuckets();
+      } else if (parentOpTraits.getCustomBucketFunctions().size() > 1) {
+        // We don't know how to merge multiple custom bucket functions. Reset bucket attributes
+        Preconditions.checkState(parentOpTraits.getBucketColNames().size() > 1);
+        listBucketCols.add(Collections.emptyList());
+        bucketFunctions.add(null);
+        numBuckets = -1;
+      } else {
+        Preconditions.checkState(parentOpTraits.getBucketColNames().size() == 1);
+        Preconditions.checkState(parentOpTraits.getCustomBucketFunctions().size() == 1);
+        final Map<String, String> inputToOutput = rs
+            .getColumnExprMap()
```

Comment:
can we place it on a same line `rs.getColumnExprMap().entrySet().stream()`?

Reply:
Done
https://github.com/apache/hive/pull/5409/commits/ae7444e8e1c2de95bfc1915af3d3c0245af4b472